index.rst

```diff
@@ -1,57 +1,451 @@
+Passing by :cpp:class:`~boost::variant` provides a convenient way to express, in code, which values are supported by :class:`~lsst.afw.typehandling.GenericMap` without committing subclasses to any particular storage mechanism.
+Unfortunately, :cpp:class:`boost::variant` handles implicitly convertible types very poorly (in particular, it gets confused when expected to convert numeric types, and cannot distinguish between ``const`` and non-``const`` versions of the same type).
+This internal detail has consequences for the public API: :class:`~lsst.afw.typehandling.GenericMap` does not currently support ``const`` values for any type except ``shared_ptr<Storable>``, and operations involving primitive types must match them exactly.
+We hope to lift these restrictions once we can migrate to :cpp:class:`std::variant` in C++17.
+
+After the nature of :class:`~lsst.afw.typehandling.GenericMap`'s type handling, the largest remaining problem was how to handle shared pointers, which are used extensively by :class:`~lsst.afw.image.ExposureInfo`.
+In keeping with C++ idioms, and in order to correctly handle polymorphism of :class:`~lsst.afw.typehandling.Storable`, :class:`~lsst.afw.typehandling.GenericMap` returns most values by reference.
+However, because the implementation holds pointers as ``shared_ptr<Storable>`` yet must return them as pointers of the correct type, its accessors create a new pointer of the desired type, which cannot be returned by reference.
+
+We chose to return smart pointers alone by value, though the inconsistency with other value types makes it much harder to write type-agnostic code against :class:`~lsst.afw.typehandling.GenericMap`.
+The alternatives were to return everything by value, which would make it impossible to support non-smart-pointer storage of :class:`~lsst.afw.typehandling.Storable`, or to always return shared pointers as ``shared_ptr<Storable>``, which would force users to perform unsafe casts in their code.
+
+It is not practical to design an idiomatic C++ API for iterating over a :class:`~lsst.afw.typehandling.GenericMap`.
+Instead, we developed a system similar to the visitors used by :cpp:class:`boost::variant` and :cpp:class:`std::variant`, where the user represents the body of the loop by a callable object that accepts values of any supported type.
+In practice the callables are usually private classes with templates or overloaded methods, but in rare cases a generic lambda can be used as well.
+This approach involves considerable boilerplate, but is more natural to users than an API written in terms of :cpp:class:`~std::variant` or some kind of iterator-like proxy.
+
+Python
+^^^^^^
+
+In Python, :class:`~lsst.afw.typehandling.GenericMap` follows the :class:`~collections.abc.Mapping` API almost exactly, aside from the need for homogeneous keys and the specific set of value types imposed by the C++ implementation.
```

Comment:
Hmm... I guess that might be true from the Python point of view. [`collections.Counter`](https://docs.python.org/3/library/collections.html#collections.Counter) is described as a `MutableMapping`, but not only does it store only integers, it returns `0` instead of `KeyError` for unknown keys. I'm pretty sure that violates Liskov substitution with `dict`. :unamused: 

I think I'll keep my wording, if only for the sake of rigor.